### ISSUE TITLE
Unable to build rpclib for android

### ISSUE THREAD

**Issue by vaasudEva** [rpclib/rpclib#133]
Hello rpclib team,
How can I build rpclib for Android ?
I am following this: https://kvurd.com/blog/compiling-a-cpp-library-for-android-with-android-studio/
But, where should I make changes in rpclib CMakeLists.txt ?

Thanks

**Comment by sztomi** [comment]
Hi @vaasudEva,

Android is unfortunately not supported, rpclib only targets Windows, Linux, MacOS. That said, it might be possible to compile it, considering that it's standards-compliant C++11 or 14, but you are on your own. 

A quick google search turns up viable alternatives: https://github.com/komamitsu/msgpack-rpc-android-example (note that you can connect to a rpclib server with any msgpack-rpc client, so you can use it on PC while using a different lib on Android, if you have a similar setup in mind).

**Comment by vaasudEva** [comment]
Hi Tamas, thanks.
I did as below:

- Download & extract https://dl.google.com/android/repository/android-ndk-r12b-linux-x86_64.zip ```./build/tools/make_standalone_toolchain.py  --arch=arm --api 24 --stl libc++ -v --install-dir=/home/me/my-android-cpp```
- ```export CC=/home/intel/my-android-cpp/bin/arm-linux-androideabi-clang```
- ```export CXX=/home/intel/my-android-cpp/bin/arm-linux-androideabi-clang++```
```
$ cmake --build .
Scanning dependencies of target rpc
[  7%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/dispatcher.cc.o
In file included from /home/intel/rpclib/lib/rpc/dispatcher.cc:1:
In file included from /home/intel/rpclib/include/rpc/dispatcher.h:11:
/home/intel/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 15%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/server.cc.o
In file included from /home/intel/rpclib/lib/rpc/server.cc:1:
In file included from /home/intel/rpclib/include/rpc/server.h:6:
/home/intel/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 23%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/client.cc.o
In file included from /home/intel/rpclib/lib/rpc/client.cc:1:
In file included from /home/intel/rpclib/include/rpc/client.h:6:
/home/intel/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 30%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/this_handler.cc.o
In file included from /home/intel/rpclib/lib/rpc/this_handler.cc:1:
In file included from /home/intel/rpclib/include/rpc/this_handler.h:8:
/home/intel/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 38%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/this_session.cc.o
In file included from /home/intel/rpclib/lib/rpc/this_session.cc:1:
In file included from /home/intel/rpclib/include/rpc/this_session.h:7:
/home/intel/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 46%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/this_server.cc.o
[ 53%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/rpc_error.cc.o
[ 61%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/detail/server_session.cc.o
In file included from /home/intel/rpclib/lib/rpc/detail/server_session.cc:1:
In file included from /home/intel/rpclib/include/rpc/detail/server_session.h:10:
/home/intel/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 69%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/detail/response.cc.o
In file included from /home/intel/rpclib/lib/rpc/detail/response.cc:1:
In file included from /home/intel/rpclib/include/rpc/detail/response.h:7:
/home/intel/rpclib/include/rpc/detail/make_unique.h:13:5: warning: 'RPCLIB_CXX_STANDARD' is not defined, evaluates to 0 [-Wundef]
#if RPCLIB_CXX_STANDARD >= 14
    ^
1 warning generated.
[ 76%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/detail/client_error.cc.o
[ 84%] Building CXX object CMakeFiles/rpc.dir/dependencies/src/format.cc.o
[ 92%] Building CXX object CMakeFiles/rpc.dir/dependencies/src/posix.cc.o
[100%] Linking CXX static library output/lib/librpc.a
[100%] Built target rpc

$ file  CMakeFiles/rpc.dir/lib/rpc/*.o
CMakeFiles/rpc.dir/lib/rpc/client.cc.o:       ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
CMakeFiles/rpc.dir/lib/rpc/dispatcher.cc.o:   ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
CMakeFiles/rpc.dir/lib/rpc/rpc_error.cc.o:    ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
CMakeFiles/rpc.dir/lib/rpc/server.cc.o:       ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
CMakeFiles/rpc.dir/lib/rpc/this_handler.cc.o: ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
CMakeFiles/rpc.dir/lib/rpc/this_server.cc.o:  ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
CMakeFiles/rpc.dir/lib/rpc/this_session.cc.o: ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped


```
Can I consider this as successful rpclib build for Android ?

**Comment by sztomi** [comment]
You could try 

    cmake -DRPCLIB_BUILD_TESTS=ON

And then run the unit tests (they are in build/output/bin/rpc_test).

Also, could you try the `dev` branch? Those embarrassing warnings should be gone from there.

**Comment by vaasudEva** [comment]
```
me@me-B85M-D3H:~/rpclib$ mkdir build && cd build
me@me-B85M-D3H:~/rpclib/build$ cmake -DRPCLIB_BUILD_TESTS=ON ..
-- The C compiler identification is Clang 3.8.256229
-- The CXX compiler identification is Clang 3.8.256229
-- Check for working C compiler: /home/me/my-android-cpp/bin/arm-linux-androideabi-clang
-- Check for working C compiler: /home/me/my-android-cpp/bin/arm-linux-androideabi-clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /home/me/my-android-cpp/bin/arm-linux-androideabi-clang++
-- Check for working CXX compiler: /home/me/my-android-cpp/bin/arm-linux-androideabi-clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
--  Clang
--  -Wall -pedantic -Weverything -Wno-c++98-compat            -Wno-c++98-compat-pedantic -Wno-padded -Wno-missing-prototypes            -pthread -std=c++11
--  rpc
--  Clang
--  -Wall -pedantic -Weverything -Wno-c++98-compat            -Wno-c++98-compat-pedantic -Wno-padded -Wno-missing-prototypes            -pthread -std=c++11
--  rpc_test
-- Configuring done
-- Generating done
-- Build files have been written to: /home/me/rpclib/build
me@me-B85M-D3H:~/rpclib/build$ cmake --build .
Scanning dependencies of target rpc
[  4%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/dispatcher.cc.o
In file included from /home/me/rpclib/lib/rpc/dispatcher.cc:1:
In file included from /home/me/rpclib/include/rpc/dispatcher.h:11:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[  8%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/server.cc.o
In file included from /home/me/rpclib/lib/rpc/server.cc:1:
In file included from /home/me/rpclib/include/rpc/server.h:6:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 12%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/client.cc.o
In file included from /home/me/rpclib/lib/rpc/client.cc:1:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:155:5: warning: 'optional_HAVE_CONSTEXPR_14' is not defined, evaluates to 0 [-Wundef]
#if optional_HAVE_CONSTEXPR_14
    ^
/home/me/rpclib/include/rpc/nonstd/optional.hpp:197:7: warning: 'nonstd_lite_HAVE_IN_PLACE_TYPES' is not defined, evaluates to 0 [-Wundef]
#if ! nonstd_lite_HAVE_IN_PLACE_TYPES
      ^
/home/me/rpclib/include/rpc/nonstd/optional.hpp:251:5: warning: 'variant_HAVE_CONDITIONAL' is not defined, evaluates to 0 [-Wundef]
#if variant_HAVE_CONDITIONAL
    ^
/home/me/rpclib/include/rpc/nonstd/optional.hpp:734:9: warning: use of this statement in a constexpr function is a C++14 extension [-Wc++14-extensions]
        assert( has_value() );
        ^
/home/me/my-android-cpp/bin/../sysroot/usr/include/assert.h:56:21: note: expanded from macro 'assert'
#  define       assert(e)       ((e) ? (void)0 : __assert2(__FILE__, __LINE__, __func__, #e))
                                ^
In file included from /home/me/rpclib/lib/rpc/client.cc:1:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:1070:1: warning: 'hash' defined as a class template here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:3065:22: note: did you mean class here?
template <class _Tp> struct hash;
                     ^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__functional_base:41:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY hash;
                     ^
In file included from /home/me/rpclib/lib/rpc/client.cc:1:
In file included from /home/me/rpclib/include/rpc/client.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
In file included from /home/me/rpclib/lib/rpc/client.cc:1:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:561:7: warning: 'bad_optional_access' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit
      [-Wweak-vtables]
class bad_optional_access : public std::logic_error
      ^
7 warnings generated.
[ 16%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/this_handler.cc.o
In file included from /home/me/rpclib/lib/rpc/this_handler.cc:1:
In file included from /home/me/rpclib/include/rpc/this_handler.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 20%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/this_session.cc.o
In file included from /home/me/rpclib/lib/rpc/this_session.cc:1:
In file included from /home/me/rpclib/include/rpc/this_session.h:7:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 25%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/this_server.cc.o
[ 29%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/rpc_error.cc.o
[ 33%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/detail/server_session.cc.o
In file included from /home/me/rpclib/lib/rpc/detail/server_session.cc:1:
In file included from /home/me/rpclib/include/rpc/detail/server_session.h:10:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 37%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/detail/response.cc.o
In file included from /home/me/rpclib/lib/rpc/detail/response.cc:1:
In file included from /home/me/rpclib/include/rpc/detail/response.h:7:
/home/me/rpclib/include/rpc/detail/make_unique.h:13:5: warning: 'RPCLIB_CXX_STANDARD' is not defined, evaluates to 0 [-Wundef]
#if RPCLIB_CXX_STANDARD >= 14
    ^
1 warning generated.
[ 41%] Building CXX object CMakeFiles/rpc.dir/lib/rpc/detail/client_error.cc.o
[ 45%] Building CXX object CMakeFiles/rpc.dir/dependencies/src/format.cc.o
[ 50%] Building CXX object CMakeFiles/rpc.dir/dependencies/src/posix.cc.o
[ 54%] Linking CXX static library output/lib/librpc.a
[ 54%] Built target rpc
Scanning dependencies of target rpc_test
[ 58%] Building CXX object CMakeFiles/rpc_test.dir/dependencies/src/gmock-gtest-all.cc.o
[ 62%] Building CXX object CMakeFiles/rpc_test.dir/tests/testmain.cc.o
[ 66%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/dispatcher_test.cc.o
In file included from /home/me/rpclib/tests/rpc/dispatcher_test.cc:5:
In file included from /home/me/rpclib/include/rpc/dispatcher.h:11:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
1 warning generated.
[ 70%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/client_test.cc.o
In file included from /home/me/rpclib/tests/rpc/client_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:734:9: warning: use of this statement in a constexpr function is a C++14 extension [-Wc++14-extensions]
        assert( has_value() );
        ^
/home/me/my-android-cpp/bin/../sysroot/usr/include/assert.h:56:21: note: expanded from macro 'assert'
#  define       assert(e)       ((e) ? (void)0 : __assert2(__FILE__, __LINE__, __func__, #e))
                                ^
In file included from /home/me/rpclib/tests/rpc/client_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:1070:1: warning: 'hash' defined as a class template here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:3065:22: note: did you mean class here?
template <class _Tp> struct hash;
                     ^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__functional_base:41:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY hash;
                     ^
In file included from /home/me/rpclib/tests/rpc/client_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
/home/me/rpclib/tests/rpc/client_test.cc:103:2: warning: no newline at end of file [-Wnewline-eof]
}
 ^
4 warnings generated.
[ 75%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/response_test.cc.o
[ 79%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/server_test.cc.o
In file included from /home/me/rpclib/tests/rpc/server_test.cc:7:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:734:9: warning: use of this statement in a constexpr function is a C++14 extension [-Wc++14-extensions]
        assert( has_value() );
        ^
/home/me/my-android-cpp/bin/../sysroot/usr/include/assert.h:56:21: note: expanded from macro 'assert'
#  define       assert(e)       ((e) ? (void)0 : __assert2(__FILE__, __LINE__, __func__, #e))
                                ^
In file included from /home/me/rpclib/tests/rpc/server_test.cc:7:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:1070:1: warning: 'hash' defined as a class template here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:3065:22: note: did you mean class here?
template <class _Tp> struct hash;
                     ^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__functional_base:41:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY hash;
                     ^
In file included from /home/me/rpclib/tests/rpc/server_test.cc:7:
In file included from /home/me/rpclib/include/rpc/client.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
3 warnings generated.
[ 83%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/this_handler_test.cc.o
In file included from /home/me/rpclib/tests/rpc/this_handler_test.cc:5:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:734:9: warning: use of this statement in a constexpr function is a C++14 extension [-Wc++14-extensions]
        assert( has_value() );
        ^
/home/me/my-android-cpp/bin/../sysroot/usr/include/assert.h:56:21: note: expanded from macro 'assert'
#  define       assert(e)       ((e) ? (void)0 : __assert2(__FILE__, __LINE__, __func__, #e))
                                ^
In file included from /home/me/rpclib/tests/rpc/this_handler_test.cc:5:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:1070:1: warning: 'hash' defined as a class template here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:3065:22: note: did you mean class here?
template <class _Tp> struct hash;
                     ^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__functional_base:41:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY hash;
                     ^
In file included from /home/me/rpclib/tests/rpc/this_handler_test.cc:5:
In file included from /home/me/rpclib/include/rpc/client.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
3 warnings generated.
[ 87%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/this_session_test.cc.o
In file included from /home/me/rpclib/tests/rpc/this_session_test.cc:6:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:734:9: warning: use of this statement in a constexpr function is a C++14 extension [-Wc++14-extensions]
        assert( has_value() );
        ^
/home/me/my-android-cpp/bin/../sysroot/usr/include/assert.h:56:21: note: expanded from macro 'assert'
#  define       assert(e)       ((e) ? (void)0 : __assert2(__FILE__, __LINE__, __func__, #e))
                                ^
In file included from /home/me/rpclib/tests/rpc/this_session_test.cc:6:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:1070:1: warning: 'hash' defined as a class template here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:3065:22: note: did you mean class here?
template <class _Tp> struct hash;
                     ^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__functional_base:41:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY hash;
                     ^
In file included from /home/me/rpclib/tests/rpc/this_session_test.cc:6:
In file included from /home/me/rpclib/include/rpc/client.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
3 warnings generated.
[ 91%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/server_session_test.cc.o
In file included from /home/me/rpclib/tests/rpc/server_session_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:734:9: warning: use of this statement in a constexpr function is a C++14 extension [-Wc++14-extensions]
        assert( has_value() );
        ^
/home/me/my-android-cpp/bin/../sysroot/usr/include/assert.h:56:21: note: expanded from macro 'assert'
#  define       assert(e)       ((e) ? (void)0 : __assert2(__FILE__, __LINE__, __func__, #e))
                                ^
In file included from /home/me/rpclib/tests/rpc/server_session_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:1070:1: warning: 'hash' defined as a class template here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:3065:22: note: did you mean class here?
template <class _Tp> struct hash;
                     ^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__functional_base:41:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY hash;
                     ^
In file included from /home/me/rpclib/tests/rpc/server_session_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
3 warnings generated.
[ 95%] Building CXX object CMakeFiles/rpc_test.dir/tests/rpc/this_server_test.cc.o
In file included from /home/me/rpclib/tests/rpc/this_server_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:734:9: warning: use of this statement in a constexpr function is a C++14 extension [-Wc++14-extensions]
        assert( has_value() );
        ^
/home/me/my-android-cpp/bin/../sysroot/usr/include/assert.h:56:21: note: expanded from macro 'assert'
#  define       assert(e)       ((e) ? (void)0 : __assert2(__FILE__, __LINE__, __func__, #e))
                                ^
In file included from /home/me/rpclib/tests/rpc/this_server_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:6:
/home/me/rpclib/include/rpc/nonstd/optional.hpp:1070:1: warning: 'hash' defined as a class template here but previously declared as a struct template [-Wmismatched-tags]
class hash< nonstd::optional<T> >
^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/memory:3065:22: note: did you mean class here?
template <class _Tp> struct hash;
                     ^
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/__functional_base:41:22: note: did you mean class here?
template <class _Tp> struct _LIBCPP_TYPE_VIS_ONLY hash;
                     ^
In file included from /home/me/rpclib/tests/rpc/this_server_test.cc:3:
In file included from /home/me/rpclib/include/rpc/client.h:8:
/home/me/rpclib/include/rpc/config.h:14:11: warning: empty paragraph passed to '\brief' command [-Wdocumentation]
//! \brief
    ~~~~~~^
3 warnings generated.
[100%] Linking CXX executable output/bin/rpc_test
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpthread
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpthread
/home/me/my-android-cpp/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lpthread
output/lib/librpc.a(this_handler.cc.o):/home/me/rpclib/lib/rpc/this_handler.cc:function rpc::this_handler(): error: undefined reference to '__cxa_thread_atexit'
clang38++: error: linker command failed with exit code 1 (use -v to see invocation)
CMakeFiles/rpc_test.dir/build.make:329: recipe for target 'output/bin/rpc_test' failed
make[2]: *** [output/bin/rpc_test] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/rpc_test.dir/all' failed
make[1]: *** [CMakeFiles/rpc_test.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
me@me-B85M-D3H:~/rpclib/build$ 
me@me-B85M-D3H:~/rpclib/build$ git branch
* dev
  master
```
Tests are failing! 
Btw, I tried building on ``dev`` branch. But I get those warning even after it.

**Comment by sztomi** [comment]
OK, it seems that this compiles with a very high warning level. I'll review this on dev when I have the chance, though most warnings seems harmless at a cursory look. On to your linker issue:

    ld: error: cannot find -lpthread

Means that your toolchain does not find libpthread. It seems that android does provide the pthread functionality, but not a libpthread. So you'll have to modify the cmake script to not use this flag when compiling for Android.

    rpc::this_handler(): error: undefined reference to '__cxa_thread_atexit'

This seems to be caused because of the standard library implementation you are compiling against. A possible workaround: https://stackoverflow.com/a/30437761/140367

**Comment by sztomi** [comment]
You were right, the warnings were indeed there on the dev branch. I have fixed them (see #132).

**Comment by vaasudEva** [comment]
Thanks Tamas.. 

Any idea why I get 

Scanning dependencies of target AirSim-rpclib
[  1%] Building CXX object rpclib/CMakeFiles/AirSim-rpclib.dir/lib/rpc/dispatcher.cc.o
In file included from /home/intel/drones-dev-tools/external/rpclib/lib/rpc/dispatcher.cc:1:
/home/intel/drones-dev-tools/external/rpclib/include/rpc/dispatcher.h:6:10: fatal error: 'atomic' file not found
#include <atomic>

when I am building rpclib as part of another project AirSim ? while I can build it successfully in independent folder!
